// vant/es/goods-action-button/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _mergeJSXProps from "@vue/babel-helper-vue-jsx-merge-props";
import { use } from '../utils';
import Button from '../button';
import { emit, inherit } from '../utils/functional';
import { functionalRoute, routeProps } from '../utils/router'; // Types

var _use = use('goods-action-button'),
    sfc = _use[0],
    bem = _use[1];

function GoodsActionButton(h, props, slots, ctx) {
  function onClick(event) {
    emit(ctx, 'click', event);
    functionalRoute(ctx);
  }

  return h(Button, _mergeJSXProps([{
    "attrs": {
      "square": true,
      "size": "large",
      "type": props.type,
      "loading": props.loading,
      "disabled": props.disabled
    },
    "class": bem(),
    "on": {
      "click": onClick
    }
  }, inherit(ctx)]), [slots.default ? slots.default() : props.text]);
}

GoodsActionButton.props = _extends({}, routeProps, {
  type: String,
  text: String,
  loading: Boolean,
  disabled: Boolean
});
export default sfc(GoodsActionButton);
